cli/ops/testing.rs

```diff
@@ -202,3 +208,192 @@ fn op_dispatch_test_event(
   Ok(())
 }
+
+enum TestOpSanitizerState {
+  None,
+  Collecting {
+    test_id: usize,
+    metrics: Vec<OpMetrics>,
+  },
+  Finished {
+    test_id: usize,
+    report: Vec<TestOpSanitizerReport>,
+  },
+}
+
+fn try_collect_metrics(
+  state: &OpState,
+  force: bool,
+  op_id_host_recv_msg: usize,
+  op_id_host_recv_ctrl: usize,
+) -> Result<Ref<Vec<OpMetrics>>, bool> {
+  let metrics = state.tracker.per_op();
+  for op_metric in &*metrics {
+    let has_pending_ops = op_metric.ops_dispatched_async
+      + op_metric.ops_dispatched_async_unref
+      > op_metric.ops_completed_async + op_metric.ops_completed_async_unref;
+    if has_pending_ops && !force {
+      let host_recv_msg = metrics
+        .get(op_id_host_recv_msg)
+        .map(|op_metric| {
+          op_metric.ops_dispatched_async + op_metric.ops_dispatched_async_unref
+            > op_metric.ops_completed_async
+              + op_metric.ops_completed_async_unref
+        })
+        .unwrap_or(false);
+      let host_recv_ctrl = metrics
+        .get(op_id_host_recv_ctrl)
+        .map(|op_metric| {
+          op_metric.ops_dispatched_async + op_metric.ops_dispatched_async_unref
+            > op_metric.ops_completed_async
+              + op_metric.ops_completed_async_unref
+        })
+        .unwrap_or(false);
+      return Err(host_recv_msg || host_recv_ctrl);
+    }
+  }
+  Ok(metrics)
+}
+
+#[op2(fast)]
+#[smi]
+// Returns:
+// 0 - success
+// 1 - for more accurate results, spin event loop and call again with force=true
+// 2 - for more accurate results, delay(1ms) and call again with force=true
+fn op_test_op_sanitizer_collect(
+  state: &mut OpState,
+  #[smi] id: usize,
+  force: bool,
+  #[smi] op_id_host_recv_msg: usize,
+  #[smi] op_id_host_recv_ctrl: usize,
+) -> Result<u8, AnyError> {
+  let metrics = {
+    let metrics = match try_collect_metrics(
+      state,
+      force,
+      op_id_host_recv_msg,
+      op_id_host_recv_ctrl,
+    ) {
+      Ok(metrics) => metrics,
+      Err(true) => {
+        return Ok(1);
+      }
+      Err(false) => {
+        return Ok(2);
+      }
+    };
+    metrics.clone()
+  };
+  let op_sanitizer_state = state.borrow_mut::<TestOpSanitizerState>();
+  if !matches!(op_sanitizer_state, TestOpSanitizerState::None) {
+    return Err(generic_error("Test metrics already being collected"));
+  }
+  *op_sanitizer_state = TestOpSanitizerState::Collecting {
+    test_id: id,
+    metrics,
+  };
+  Ok(0)
+}
+
+#[derive(Serialize)]
+#[serde(rename_all = "camelCase")]
+struct TestOpSanitizerReport {
+  id: usize,
+  diff: i64,
+}
+
+#[op2(fast)]
+#[smi]
+// Returns:
+// 0 - sanitizer finished with no pending ops
+// 1 - for more accurate results, spin event loop and call again with force=true
+// 2 - for more accurate results, delay(1ms) and call again with force=true
+// 3 - sanitizer finished with pending ops, collect the report with op_test_op_sanitizer_report
+fn op_test_op_sanitizer_finish(
+  state: &mut OpState,
+  #[smi] id: usize,
```

Comment:
For `#[smi]`-tagged values I'd suggest using a typedef and a smaller-sized integer. You can also use untagged `u32` here without penalty and it'll take the fastest path.